### PR TITLE
Fix memory region api

### LINKED ISSUE #6547: Memory region APIs are unsound
**Describe the bug**
The memory region APIs are not marked `unsafe`, but they take arbitrary addresses (represented as `u64`).  The addresses are used so this is unsound.  Furthermore, the fields of `UserMemoryRegion` are all public, so one can create an invalid `UserMemoryRegion` in safe code.

**To Reproduce**
Call the memory region APIs with junk `userspace_addr` parameters.

**Version**
Latest git main.

**VM configuration**
Not applicable.  This is a problem with Cloud Hypervisor’s internal APIs.

**Logs**
Not applicable.  This is a problem with Cloud Hypervisor’s internal APIs.

### PR DESCRIPTION
**Pull request by Pivnoy**
Fix #6547

Make private fields of ```UserMemoryRegin```.
Mark ```make_user_memory_region``` as ```unsafe```, add relevant comments

### CODE REVIEW

File: hypervisor/src/kvm/mod.rs [hunk 1]
```diff
@@ -602,10 +602,9 @@ impl vm::Vm for KvmVm {
             .map_err(|e| vm::HypervisorVmError::SetGsiRouting(e.into()))
     }
 
-    ///
+    /// # Safety
     /// Creates a memory region structure that can be used with {create/remove}_user_memory_region
```
Comment (alyssais) [inline hypervisor/src/kvm/mod.rs:606]: A safety doc comment is supposed to explain the invariants the function requires users to ensure — not just to explain what the function does, which should go above the safety comment.

**Review by alyssais — CHANGES_REQUESTED** [review]

**Comment by alyssais** [comment]
I'm not familiar with this API, so I might have missed something, but why does the caller need to guarantee the validity of `slot`?  What's unsafe about passing an invalid slot and getting an error back?  I'm also not sure about guest_phys_addr — it's not immediately clear to me that that would be unsafe in Rust terms either.

Ideally the safety comment would describe what specifically the caller needs to guarantee, rather than just "validity" — maybe a maintainer could help explain those invariants if you're not sure.

**Comment by rbradford** [comment]
Thank you for looking at this change - i'm not really sure this is particularly helpful. Labelling this method as unsafe doesn't help since there is no "safe" way to do this in lieu. Instead it just requires adding extra unsafe declarations around the code making it unnecessarily messy. The use of unsafe in API is to make sure the caller takes extra steps (i.e. to force the developer to take time and double check their work) however I don't think that is useful in this case.

**Comment by alyssais** [comment]
> Thank you for looking at this change - i'm not really sure this is particularly helpful. Labelling this method as unsafe doesn't help since there is no "safe" way to do this in lieu. Instead it just requires adding extra unsafe declarations around the code making it unnecessarily messy. The use of unsafe in API is to make sure the caller takes extra steps (i.e. to force the developer to take time and double check their work) however I don't think that is useful in this case.

I strongly disagree that Cloud Hypervisor should contain methods that are known to be unsafe, but not marked as such.  `unsafe` has a clear meaning that should be applied consistently across different Rust codebases — this is critical for Rust developers to be able to produce sound code while moving between different projects.  It's common for there to be no safe equivalent to unsafe functions.

`unsafe` is not just useful when writing code, it's also useful when debugging it, and that's a situation where it's very likely that the person doing the debugging is not familiar with the code enough to know about a deliberate convention of not marking functions known to be unsafe.  One of the soundness issues (#6647) I found in Cloud Hypervisor recently was due to incorrect use of a function that was known to have preconditions, but not marked as unsafe.  If the rules of `unsafe` had been correctly followed, that issue couldn't have happened.

**Comment by rbradford** [comment]
> > Thank you for looking at this change - i'm not really sure this is particularly helpful. Labelling this method as unsafe doesn't help since there is no "safe" way to do this in lieu. Instead it just requires adding extra unsafe declarations around the code making it unnecessarily messy. The use of unsafe in API is to make sure the caller takes extra steps (i.e. to force the developer to take time and double check their work) however I don't think that is useful in this case.
> 
> I strongly disagree that Cloud Hypervisor should contain methods that are known to be unsafe, but not marked as such. `unsafe` has a clear meaning that should be applied consistently across different Rust codebases — this is critical for Rust developers to be able to produce sound code while moving between different projects. It's common for there to be no safe equivalent to unsafe functions.
> 
> `unsafe` is not just useful when writing code, it's also useful when debugging it, and that's a situation where it's very likely that the person doing the debugging is not familiar with the code enough to know about a deliberate convention of not marking functions known to be unsafe. One of the soundness issues (#6647) I found in Cloud Hypervisor recently was due to incorrect use of a function that was known to have preconditions, but not marked as unsafe. If the rules of `unsafe` had been correctly followed, that issue couldn't have happened.

Okay - I agree that there are probably bugs in the code due to mistakes with assumptions but i'm not 100% sure that labelling a function that just populates a struct as unsafe is useful. That function itself can't have incorrect side effects since all it does (at least on KVM) is copy the values in.

**Comment by alyssais** [comment]
> Okay - I agree that there are probably bugs in the code due to mistakes with assumptions but i'm not 100% sure that labelling a function that just populates a struct as unsafe is useful.

Well, it's our choice which APIs to mark as unsafe.  The APIs that consume the struct could be marked as unsafe instead, but it's very common for Rust abstractions to provide a wrapper type that is unsafe to construct with arbitrary values, so that the methods of that type can be safe, reducing the number of `unsafe` blocks required overall.

**Comment by liuw** [comment]
Are there anything we need to do regarding https://github.com/cloud-hypervisor/cloud-hypervisor/issues/6547?

Are people happy with leaving things as they are?

**Comment by rbradford** [comment]
I think the PR just needs a rebase.

**Pull request closed without merging**